librad/src/id/entity.rs

```diff
@@ -0,0 +1,515 @@
+                        Signatory::User(uri) => Some(uri.to_string()),
+                        Signatory::OwnedKey => None,
+                    },
+                    sig: s.sig.to_bs58(),
+                },
+            );
+        }
+
+        let keys = HashSet::from_iter(self.keys().iter().map(|k| k.to_bs58()));
+        let certifiers = HashSet::from_iter(self.certifiers().iter().map(|c| c.to_string()));
+
+        data::EntityData {
+            name: Some(self.name.to_owned()),
+            revision: Some(self.revision),
+            hash: Some(
+                bs58::encode(&self.hash)
+                    .with_alphabet(bs58::alphabet::BITCOIN)
+                    .into_string(),
+            ),
+            parent_hash: self.parent_hash.to_owned().map(|h| {
+                bs58::encode(h)
+                    .with_alphabet(bs58::alphabet::BITCOIN)
+                    .into_string()
+            }),
+            signatures: Some(signatures),
+            keys,
+            certifiers,
+            info: self.info.to_owned(),
+        }
+    }
+
+    pub fn to_builder(&self) -> data::EntityData<T> {
+        self.to_data().clear_hash().clear_signatures()
+    }
+
+    pub fn hash(&self) -> &Multihash {
+        &self.hash
+    }
+    pub fn uri(&self) -> RadicleUri {
+        RadicleUri::new(self.hash.to_owned())
+    }
+
+    pub fn signatures(&self) -> &HashMap<PublicKey, EntitySignature> {
+        &self.signatures
+    }
+
+    pub fn keys(&self) -> &HashSet<PublicKey> {
+        &self.keys
+    }
+    fn keys_count(&self) -> usize {
+        self.keys.len()
+    }
+    fn has_key(&self, key: &PublicKey) -> bool {
+        self.keys.contains(key)
+    }
+
+    pub fn certifiers(&self) -> &HashSet<RadicleUri> {
+        &self.certifiers
+    }
+    fn certifiers_count(&self) -> usize {
+        self.certifiers.len()
+    }
+    fn has_certifier(&self, c: &RadicleUri) -> bool {
+        self.certifiers.contains(c)
+    }
+
+    pub fn canonical_data(&self) -> Result<Vec<u8>, Error> {
+        Ok(self.to_data().canonical_data()?)
+    }
+    pub fn to_json_writer<W>(&self, writer: W) -> Result<(), Error>
+    where
+        W: std::io::Write,
+    {
+        self.to_data().to_json_writer(writer)?;
+        Ok(())
+    }
+    pub fn to_json_string(&self) -> Result<String, Error> {
+        Ok(self.to_data().to_json_string()?)
+    }
+
+    pub fn from_json_reader<R>(r: R) -> Result<Self, Error>
+    where
+        R: std::io::Read,
+    {
+        Ok(Self::from_data(
+            data::EntityData::from_json_reader(r)?.to_owned(),
+        )?)
+    }
+    pub fn from_json_str(s: &str) -> Result<Self, Error> {
+        Ok(Self::from_data(
+            data::EntityData::from_json_str(s)?.to_owned(),
+        )?)
+    }
+
+    pub fn compute_hash(&self) -> Result<Multihash, Error> {
+        Ok(Sha2_256::digest(&self.canonical_data()?))
+    }
+
+    pub async fn check_user_key(
+        &self,
+        user_uri: &RadicleUri,
```

Comment:
I think it might make more sense to write this function in terms of `User` and then build the `RadicleUri` and `Resolver<User>` function on top of that. It separates the validation logic from the async logic.

Reply:
Bah, that function was actually dead code, a leftover from the previous approach, now both unused and wrong.
I removed it.